Multiple idps (#205)
mulesoft/docs-api-manager: modules/ROOT/pages/configure-multiple-credential-providers.adoc

= Configuring Multiple Identity Providers for Client Management
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: credential providers, external client providers

> You can use multiple client providers, to help you enforce security and regulations in your business organization. These client providers, such as, OpenAM and PingFederate, enable you to secure your operational data, such as client credentials and access token. 

Reviewer: ```suggestion
You can use multiple client providers to help you enforce security and regulations in your business organization. These client providers, such as OpenAM and PingFederate, enable you to secure your operational data, such as client credentials and access tokens. 
```

Author: Done! Thanks!

> You can use an Anypoint Platform client provider (native client provider) or implement an external client provider. If you do not specify a client provider for your environment or business group, Anypoint Platform uses its native client provider by default.

Reviewer: Why provide both names?
```suggestion
You can use an Anypoint Platform native client provider (default) or implement an external client provider. 
```

Reviewer 2: istead of *implement*, I would say something more similar to *associate*

Author: I said configure..:-)

> To assign separate client providers for different organizations and environments, you must first enable clients applications to be authenticated using OAuth for the client providers that you want to implement in your organization.

Reviewer: ```suggestion
To assign separate client providers for different organizations and environments, you must first enable client applications to be authenticated using OAuth for the client providers that you want to implement in your organization.
```

Author: Done! Thanks!

> When you use multiple client providers, you can use them to:

Reviewer: ```suggestion
You use multiple client providers for:
```

Author: Thanks! Done!

* Different business groups
* Different environments, such as production, QA, and sandbox
* Internal and external APIs

== How This Works

//Need help here

> When you enable external client providers, they act as authorization servers to issue access tokens to a client application for accessing an API. When you request access to a client application, the corresponding client ID and client secret credentials are created in the IdP specified in that API specification.

Reviewer: GLOBALLY work to make sentences shorter, simpler, and more clear, by avoiding restatement and unnecessary modifiers. Let the sentences in a para work together to build the point, rather than putting all of it in one.
```suggestion
When you enable external client providers, they issue tokens that enable a client application to access an API. When you request access to that client application, the corresponding client ID and client secret are created in the IdP specified in that API specification.
```

Author: Done! Thanks!

> If an API has only applied the Client ID Enforcement policy and not applied the OAuth policy, the client application is still created in an external IdP.

Reviewer: ```suggestion
If an API applies the Client ID Enforcement policy but not the OAuth policy, the client application is still created in an external IdP.
```

Author: Done! Thanks!

> When an OAuth policy is deployed to API gateway, the platform automatically injects IdP configuration, such as the token URL and token introspection url in the policy configuration. The IdP configuration is completely hidden from a user in API Manager.

Reviewer: "injects configuration"? 
```suggestion
When an OAuth policy is deployed to the API gateway, the platform automatically injects IdP configuration (such as the token URL and token introspection URL) to the policy. The IdP configuration is completely hidden from the  API Manager user.
```

Author: Done! Thanks!

If no external IdPs are enabled, Access Manager acts as the default authorization server in that case. 
//[PA:-Tomas not sure what the following statement is trying to say.]
// OOTB OAuth policies with an external IdP can’t be applied 
// ClientId enforcement policy can be applied.

> 

Reviewer 2: Answering the comment:
It means that if the customer doesn't associate an external client provider, there are a set of policies that they can't apply (those that rely on an external client provider)

Author: Ok that is already covered.

== The Use Case

> To better understand how to implement the multiple external credential provider feature, let us look at the example of "Good Weather" company that is in the business of forecasting weather for a given location and time. Good Weather collects quantitative data about the current state of the atmosphere at a given place and, using meteorology, predicts how that current state will change over time.

Reviewer: ```suggestion
To better understand how to implement the multiple external credential provider feature, imagine the "Good Weather" company, which forecasts weather for a given location and time. Good Weather collects quantitative data about the current state of the atmosphere at a given place and, using meteorology, predicts how that current state will change over time.
```
This is a pretty lengthy description that can probably be summarized as 
```suggestion
To better understand how to implement multiple external credential providers, imagine a company, "Good Weather," that uses meteorology to forecast changing weather.
```

Author: Done! Thanks!

> Good Weather has several sub-organizations running under the master organization. These sub-organizations exist based on the functions they perform, for example, Engineering and Forecasting. The Forecasting department researches and collects recent meterological data from various sources. This data is then consumed by various client applications. While the Engineering department maintains the applications that host the forecasting research data.

Reviewer: ```suggestion
The Good Weather organization has several departments. These departments each perform specific functions, such as the Engineering department and the Forecasting department. The Forecasting department collects for various client applications recent meteorological data, while the Engineering department maintains the applications that host the forecasting data.
```

Reviewer: I think you can probably omit the next paragraph and go straight to "To ensure security and segregation...."

Author: No the set up is required.

> Good Weather needs to ensure that the forecasting data is accurate at all times. For that reason the weather APIs must be accessed only by employees of the Forecasting sub-organization. Good Weather also wants to ensure that the production APIs are the single source of truth at all times.

Reviewer: I'd omit this. But if not...
```suggestion
Because Good Weather must ensure that its weather-forecasting data is accurate and secure, the weather APIs are accessed only by employees of the Forecasting department. Good Weather also wants to ensure that the production APIs are the single source of truth at all times.
```

Author: Done! Thanks!

> To ensure security and segregation of access for these different organizations, environments, and APIs, Good Weather IT sets up a separate external client providers for the Forecasting sub-organization.

Reviewer: ```suggestion
To ensure security and access segregation for different organizations, environments, and APIs, Good Weather IT sets up separate external client providers for the Forecasting department.
```

Author: Done! Thanks!

//example of how they implement the client providers.
//Tomas how will they best implement this?

== Guidelines for Implementing Multiple Client Providers

> Before you implement multiple client providers in Anypoint Platform, ensure that you carefully review the following guidelines and best practices:

Reviewer: ```suggestion
Before you implement multiple client providers in Anypoint Platform, carefully review and understand the following points:
```

Author: Done! Thanks!

* When changing the client provider for an existing API instance, ensure that the API:
** Has no contracts

> ** Has no client provider-related policies applied, such as PingFederate and OpenID policies.

Reviewer: * When you change the client provider for an existing API instance, ensure that the API has no contracts and has no client provider-related policies applied, such as PingFederate and OpenID policies.

Author: No, will keep it sub-bulleted.

* When promoting and importing an API:

> * If you assign an external client provider to an environment, all existing APIs in that environment display Anypoint Platform in the UI by default. 

Reviewer 2: I had commented on the doc. If you didn't an IDP assigned to your environment, after you assign an external IDP to it, all EXISTING APIs will still be using Anypoint. You can then reassign the required external IdP for that API.

Author: Done! Thanks!

+

> You must then reassign the appropriate external client provider for that API.

Reviewer: * When you promote and import an API and assign an external client provider to an environment, all existing APIs in that environment display Anypoint Platform in the UI by default. 
You must then reassign the appropriate external client provider for that API.

Author: These are two separate bullets and I'm waiting for info from Tomas.

> * After you assign an external client provider to an API, you cannot reassign the client provider back to the default native Anypoint Platform credential provider.

Reviewer: ```suggestion
* After you assign an external client provider to an API, you cannot return to using the default native Anypoint Platform credential provider.
```

Author: Done! Thanks!

> * If you remove a client provider from an environment, all existing APIs and client applications using that client provider continue to work.

Reviewer: ```suggestion
* If you remove a client provider from an environment, all existing APIs and client applications using that client provider continue to work.
```

Author: Couldn't get what is the change there..

> * If you delete an client provider from a master organization, all existing APIs and client applications using that client provider default to the native Anypoint Platform client provider. 

Reviewer: ```suggestion
* If you delete a client provider from a master organization, all existing APIs and client applications using that client provider default to the native Anypoint Platform client provider, and all contracts remain intact. 
```

Reviewer 2: If you delete an IdP from a master organization, all existing APIs will be assigned to Anypoint IdP. Even though contracts will remain intact, policies that authorize against the provider will start failing as its configuration has been deleted.

Author: Done! Thanks!

+

> All contracts remain intact.

Reviewer: moved

Author: Done! Thanks!

> == Configuring Multiple Client Providers User Flow

Reviewer: ```suggestion
== Configuring a User Flow with Multiple Client Providers
```

Author: No it is the other way around. You are not configuring a user flow, but this is a user flow for configuring multiple client providers.

> Based on the use case explained in the previous section, Good Weather uses the following flow to implement multiple external client providers:

Reviewer: ```suggestion
Returning to the Good Weather use case, imagine that the company uses the following flow to implement multiple external client providers:
```

Author: Done! Thanks!

> image::multiple-IdP-user-flow.png[User Flow for Configuring Multiple External Client Providers]

Reviewer 3: This diagram does not follow the guidelines. And it is in an incorrectly named folder.

Author: Fixed! Thanks!

> 

Reviewer: Should the graphic be followed by some text that sets the audience expectation for the following subtopics?

Reviewer 3: Yes, the requirement for that is in the guidelines.

Author: Done! thanks!

=== Tasks for Implementing Multiple Client Providers Based on Roles

> Based on your role in your organization and your requirements, you might need to perform the following set of tasks to implement multiple client providers:

Reviewer: Is this text transitioning away from the example that it just provided an example flow for? If so, how is this a subtopic? Clarify narrative flow between use case and general use.

Author: Done! Thanks!

> Based on your role in your organization and your requirements, you might need to perform the following set of tasks to implement multiple client providers:

Reviewer: ```suggestion
Based on your role and the needs of your organization, you might perform the following tasks to implement multiple client providers:
```

Author: Done! Thanks!

[%header,cols="20a,40a,20a,50a"]

> |===

Reviewer: Do we really want to create and define specific user titles/roles instead of focusing on the need/requirement?
Personas/user titles are not standard among companies.

Reviewer: If a table requires a Notes column, consider that a table might not be the best vehicle for the information, which should be SCANNABLE.
Could this instead be a list of user tasks or requirements, followed by location and "notes"?

Author: For me, it works the way it is represented now.

| Role | Task | Where to Perform | Notes

> | Business Group Administrator | Assign multiple IdPs to an environment and also assign a default IdP. | Access Management | Using multiple IdPs provides flexibility to choose an IdP in API configuration inside API Manager, which can be useful when multiple teams are using the same environment for building APIs, but still want to use separate IdPs. This also helps in use cases when separate API instances are created for internal and external consumers, and separate IdPs based on a consumer is required. 

Reviewer: ```suggestion
| Business Group Administrator | Assign multiple IdPs, including a default, to an environment | Access Management | Using multiple IdPs provides flexibility to choose an IdP in API configuration inside API Manager, which can be useful when multiple teams are using the same environment for building APIs, but still want to use separate IdPs. This also helps in use cases when separate API instances are created for internal and external consumers, and separate IdPs based on a consumer is required. 
```

Author: Thanks! Done!

> | Business Group Administrator | Be notified about any possible interruption to API consumers when I am about to change any IdP configuration.  | Access Management | None

Reviewer: ```suggestion
| Business Group Administrator | Receive notification about possible interruption to API consumers before changing IdP configuration.  | Access Management | None
```

Author: Done! Thanks!

> | Master organization Administrator | Change the default IdP across all environments and across all business groups. | Access Management | Resetting the default IDP updates all environments across all business groups to a selected IDP, without affecting existing client applications. This way, when a new IdP is created for the organization, all environments are updated with just one change.

Reviewer: ```suggestion
| Master Organization Administrator | Change the default IdP across all environments and business groups. | Access Management | Resetting the default IDP updates all environments across all business groups to a selected IdP, without affecting existing client applications. This way, when a new IdP is created for the organization, all environments are updated with just one change.
```

Author: Done! Thanks!

> | Master organization Administrator | Change the default IdP across all environments and across all business groups. | Access Management | Resetting the default IDP updates all environments across all business groups to a selected IDP, without affecting existing client applications. This way, when a new IdP is created for the organization, all environments are updated with just one change.

Reviewer 2: "Change the default IdP across all environments and across all business groups." This is not something that can be done AFAIK.

Author: commented out

> | API Manager Administrator | Review which client provider an application belongs to on the client administrator page to be able to differentiate between different applications with the same name. | Access Management | None

Reviewer: ```suggestion
| API Manager Administrator | Review tow which client provider an application belongs to differentiate between different applications with the same name. | Access Management | None
```

Author: Done! Thanks!

> | API Owner | Configure my API in API Manager to automatically map to a default IdP configured by the business group administrator. I can also select a different IdP, if available | API Manager | None

Reviewer: ```suggestion
| API Owner | Configure owned API in API Manager to automatically map to a default IdP configured by the business group administrator. Owner can also select a different IdP, if available. | API Manager | None
```

Author: Done! Thanks!

> | API Owner | When promoting or importing an API between environments, specify an IdP from the target environment. | API Manager | If there is just one (default) IdP specified for the target environment, it is automatically assigned to the API on promotion. 

Reviewer: Did you meant to repeat API Owner? Will the audience be confused by this; should the two entries be combined?
```suggestion
| API Owner | When promoting or importing an API between environments, specify an IdP from the target environment. | API Manager | If there is just one (default) IdP specified for the target environment, it is automatically assigned to the API on promotion. 
```
Sounds repetitive of previous.

Author: Nope! Different scenarios!

| Application User | Request access to an API based on the API instance that exists in a specific environment. | Exchange | When you select the API for which you want to request access, you must first choose the API instance to which you are requesting access. The API instance is filtered based on the environment.
|===

=== Configure Multiple Client Providers

> To configure multiple external client providers:

Reviewer: ```suggestion
The following flow to configure multiple external client providers requires the participation of all user roles.
```

Reviewer 2: Don't think this suggestion is correct.

Author: Yes I agree. So keeping the original.

> 

Reviewer: These are steps; use step format.

Author: Yes totally! How did I miss that.

* Step 1: From *Access Management*, first enable xref:access-management::managing-api-clients.adoc[client management].
+

> As an Anypoint Administrator, you must first enable clients applications to be authenticated using OAuth for the client providers that you want to implement in your organization.

Reviewer: Suggest omitting; redundant of the step above it. Otherwise...:
```suggestion
The Anypoint Administrator must first enable client applications to be authenticated using OAuth for the client providers to be implemented in the organization.
```

Author: Done! Thanks!

> * Step 2: From *Access Management*, xref:access-management::environments.adoc#add-an-idp-to-an-existing-environment[assign client providers] to environments and business groups.

Reviewer: ```suggestion
. From *Access Management*, the Anypoint Administrator xref:access-management::environments.adoc#add-an-idp-to-an-existing-environment[assigns client providers] to environments and business groups.
```

Author: Done! Thanks!

+

> As a Business Administrator, you must then configure multiple client providers and assign them to specific business organizations and environments.

Reviewer: ```suggestion
. The Business Administrator configures multiple client providers and assigns them to specific business organizations and environments.
```

Author: Done! Thanks!

> As a Business Administrator, you must then configure multiple client providers and assign them to specific business organizations and environments.

Reviewer 2: stricly speaking, this could be one client provider. We don't need to put multiple here.

Author: Done! Thanks!

> * Step 3: From *API Manager*, xref::create-instance-task.adoc[create or modify] APIs using the new IdPs.

Reviewer: ```suggestion
. The API Owner xref::create-instance-task.adoc[creates or modifies] APIs using the new IdPs.
```

Author: Does not fit. Made changes to fit the rest of the format.

+

> As an API Owner, you can create an API, import an API from Exchange or modify an API in API Manager. Optionally, you can promote an API from one environment to another, and import or export an API, as described in the user flow diagram in the previous section. 

Reviewer: ```suggestion
As an API Owner, you can create an API, import an API from Anypoint Exchange, or modify an API in API Manager. Optionally, an owner can promote an API from one environment to another, and import or export an API, as described in the user flow diagram in the previous section. 
```
Replace reference to "the user flow diagram in the previous section" with the section title.

Reviewer 2: Instead of owner I prefer to leave the `you`

Author: Made changes to match the 3rd person reference.

> * Step 4: From *API Manager*, xref:automated-policies-landing-page.adoc[configure policies] associated to the new IdP.

Reviewer: ```suggestion
. From *API Manager*, the API Manager Administrator xref:automated-policies-landing-page.adoc[configures policies] for the new IdP.
```

Author: Kept as is.

+

> As an API Manager Administrator, you can apply policies to an API based on the client provider associated to it.

Reviewer: Omit; repeats step.

Author: Removed the duplicate phrase from the step.

> * Step 5: From *Exchange*, xref:exchange::to-request-access.adoc[request access] to an API instance within a specific environment for business organization.

Reviewer: ```suggestion
. From *Exchange*, the API consumer xref:exchange::to-request-access.adoc[requests access] to an API instance within a specific environment for a business organization.
```

Author: Kept as is.

+

> As an API consumer, you can request access to an API instance from registered client applications using private exchange or public portals.

Reviewer: ```suggestion
API consumers can request access to an API instance from registered client applications using private Exchange or public portals.
```

Author: Done! Thanks!

== See Also

> * xref:manage-exchange-api-task.adoc[Creating API Instances by Importing From Exchange]

Reviewer: ```suggestion
* xref:manage-exchange-api-task.adoc[Creating API Instances by Importing from Exchange]
```

Author: Done! Thanks!

* xref:access-management::managing-api-clients.adoc[Managing Clients in Access Management]

> * xref:import-api-task.adoc[Importing APIs With Multiple Client Providers]

Reviewer: ```suggestion
* xref:import-api-task.adoc[Importing APIs with Multiple Client Providers]
```

Author: Done! Thanks!

> * xref:export-api-latest-task.adoc[Exporting APIs With Multiple Client Providers]

Reviewer: ```suggestion
* xref:export-api-latest-task.adoc[Exporting APIs with Multiple Client Providers]
```

Author: Done! Thanks!

> * xref:promote-api-task.adoc[Promoting APIs With Multiple Client Providers]

Reviewer: ```suggestion
* xref:promote-api-task.adoc[Promoting APIs with Multiple Client Providers]
```

Author: Done! Thanks!

= Configuring Multiple Credential Providers
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: credential providers, external client providers

With Anypoint Platform, you can use multiple client providers, also known as identity providers (IdP) to manage client credentials to help you implement security policies and regulations. This ensures that client applications that are created to access APIs are stored in different systems based on the environment where API is located. Separate client providers helps with segregating client application management for:

* Different business groups
* Different environments, such as production, QA, and sandbox
* Internal and external APIs

> For example, a “sandbox” client, whose credentials live in an external Okta IdP, can have contracts only with API instances in the organization “Sandbox A” environment of the business group A and “open” environment of business group B. Similarly, a “core-internal” client, whose credentials live in the private ADFS IdP of the organization, can have contracts only with API instances in the “prod” environment of the business group X. 

Reviewer 4: > “open” environment of business group

@priyaayyar Can you remove this part? It's confusing

Author: Done! Thanks!

> == How This Works

Reviewer 4: @priyaayyar @tbruno this section is really weak and lacks context. It needs more information on how configuration is shared between APIM and Access Management.

Author: Need help on this one.

When you enable external IdPs, they act as authorization servers to issue access tokens to a client application for accessing an API. When you request access to a client application, the corresponding client ID and client secret credentials are created in the IdP specified in that API specification.

If an API has only applied the Client ID Enforcement policy and not applied the OAuth policy the client application is still created in an external IdP.

> The API owner does not have access to the IdP configuration in API Manager. 

Reviewer 4: @tbruno what does this mean? 

> The API owner does not have access to the IdP configuration in API Manager.

Author: Removed it!

When an OAuth policy is deployed to API gateway, the platform automatically injects IdP configuration, such as the token URL and token introspection url in the policy configuration. The IdP configuration is completely hidden from a user in API Manager.

If no external IdPs are enabled, Access Manager acts as the default authorization server in that case. 
//[PA:-Tomas not sure what the following statement is trying to say.]
// Any client apps (clientID/clientSecret) that’re created for accessing an API are created and maintained within Anypoint.
// OOTB OAuth policies with an external IdP can’t be applied 
// ClientId enforcement policy can be applied.

> == The Use Case

Reviewer 4: @priyaayyar @tbruno Is this the correct use case? It doesn't look like a use case of client management instead it looks a use case of single-sign-on

> To better understand how to implement the multiple external credential provider feature, let us go back to our example of "Good Weather" company that is in the business of forecasting weather for a given location and time. Good Weather collects quantitative data about the current state of the atmosphere at a given place and, using meteorology, predicts how that current state will change over time.

Reviewer 4: @priyaayyar where else is this example used?

> let us go back to our example of "Good Weather"

Good Weather has several sub-organizations running under the master organization. These sub-organizations exist based on the functions they perform, for example, finance, HR, and R&D. The R&D department researches and collects recent meterological data from various sources. This data is then consumed by various client applications. 

Therefore, Good Weather needs to ensure that this data is accurate at all times. For that reason the weather APIs must be accessed only by employees of the R&D department. Good Weather also want to ensure that the production APIs and the single source of truth at all times.

On the other hand, the HR sub-organization wants to ensure that the employees in R&D do not have access to their payroll APIs. HR also wants to restrict access to their internal tax and payroll APIs.

To ensure security and segregation of access to these different organizations, environments, and APIs, Good Weather IT sets up a separate external IdP for each sub-organization and environments (production and non-production).

== Guidelines for Implementing Multiple Client Providers

> Before you implement multiple client providers in your instance of Anypoint Platform, ensure that you carefully review the following guidelines and best practices:

Reviewer 4: @priyaayyar What is an **instance** of Anypoint platform? 

> in your instance of Anypoint Platform,

Author: Removed!

> * If you assign an external IdP to an environment, all existing APIs in that environment display Anypoint Platform in the UI dropdown. 

Reviewer 4: This lacks context and difficult to follow. Please clarify this statement

Author: Done! Thanks!

+
You can then reassign the required external IdP for that API.
* After you assign a external IdP to an API, you cannot change this IdP back to be the default of Anypoint Platform as a credential provider.
* If you remove an IdP from an environment, all existing APIs and client applications using that IdP will continue working.

> * If you delete an IdP from a master organization, all existing APIs and client applications using that IdP will change to the default Anypoint Platform IdP . 

Reviewer 4: @tbruno Please verify this comment. This doesn't sound right

+
All contracts will remain intact.
* When changing the IdP for an existing API instance, ensure that:
** The API has no contracts
** The API has no IdP-related policies applied, such as PingFederate and OpenID policies.
* When promoting and importing an API:
//[PA: Tomas what are the rules? I was not able to access the LucidChart]

== Configuring External Client Providers User Flow

Based on the use case explained in the previous section, Good Weather uses the following flow to implement multiple external IdPs:

> image::multiple-IdP-user-flow.png[User Flow for Configuring Multiple External Client Providers]

Reviewer 4: the second and third node in the image duplicate?

Author: Done! Fixed!
